### DISCUSSION TITLE
Writing Testing Library tests that use certain Skeleton components

### DISCUSSION THREAD

**Discussion by is-sematell** [skeletonlabs/skeleton#2891] [Q&A] [answered]
I'm working on a project that uses SvelteKit, Skeleton, and Testing Library. What is the best way of using and testing Skeleton components like Popup or Modal in Testing Library tests? Can they be used directly, or do they all need to be mocked?

**Comment by endigo9740** [comment]
Hey @is-sematell, so two things here:

1. Make sure you only test *your work* - as in code and code you've implemented specific to your application
2. Use the right tool for the job at hand - as there's multiple ways to test

Skeleton tests our own components and features, so you do not need to. That would be redundant. [Here's our modals test for example](https://github.com/skeletonlabs/skeleton/blob/dev/packages/skeleton/src/lib/utilities/Modal/Modal.test.ts). What you should focus on is testing your own *custom modals*, which utilize a component embedded within a modal window. In that case, you would test the component (directly in isolation) just like any other, and Outside the scope of the wrapping modal system (which again is already tested on our side).

For overlay elements (popovers, modals, etc), those can be a bit harder to test with unit tests (ex: Vitest + Testing Library) since you need a trigger, you need the test to click the trigger, await the modal to display, then test the results. But that is a scenario Testing Library can fulfill. Typically paired with an async task to await the change in interface. You can see a brief example of triggering clicks here:
https://testing-library.com/docs/svelte-testing-library/example/

One technique I've used with success here is to create a small co-located component alongside the test that is purpose built to replicate a trigger + modal/popover in isolation. That's the primary goal of unit tests after all - testing as isolated as possible. That's a bit closer to integration testing, due to the extra scaffolding, but still something these tools are equipped to handle.

Finally, if you really need to test a feature as it functions in the user flow. That's where e2e testing may be a lot more valuable. You can use testing tools like [Playwright](https://playwright.dev/) for that.

As for the specifics for implementing the tests themselves, that's a bit outside the scope of our support. But I would recommend reading through the Testing Library documentation to understand how it operates. However, Skeleton is open source so you can reference our component tests here:

- v2: https://github.com/skeletonlabs/skeleton/tree/dev/packages/skeleton/src/lib/components
- v3: https://github.com/skeletonlabs/skeleton/tree/next/packages/skeleton-react/src/lib/components

Note that our v3 tests a bit better quality, as we've matured in our understanding of the testing tooling, but currently limited only to our React components. This is due to Vitest and Test Library not having thorough documentation for the Svelte 5 pre-release versions. Now that Svelte 5 has launched, we expect that to change, and of course bring parity to our new components.

Hope this helps!

**Reply by is-sematell** [reply]
Hey @endigo9740,

Thanks for the detailed answer. Just to make it clear, I don't intend on testing Skeleton components directly, that would be way out of scope. I only intend to test our own components that utilize Skeleton components under the hood.

Nonetheless, that would indeed require the Skeleton components like Modal or Popup to work with Testing Library. Then the test would have to trigger, for example, a popup, assert its contents, etc.

So, how should Skeleton be set up for Testing Library? Would I need to create some kind of wrapper that initializes the Skeleton stores? Are all components fully compatible to work seamlessly in Testing Library, or are there any technical limitations?

**Reply by is-sematell** [reply]
By the way, I don't see any test example for [Popup](https://github.com/skeletonlabs/skeleton/tree/dev/packages/skeleton/src/lib/utilities/Popup). Is there any reason for that?

**Reply by is-sematell** [reply]
@endigo9740 I've created a Tooltip component that utilizes Skeleton popup. It works just fine in the browser.

However, when I try to test it via Testing Library, I get the below error. Seems like it is caused by `userEvent.hover`. 

Am I missing something, or is this a bug?

## Error

```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Uncaught Exception ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
SyntaxError: 'a[href], button, input, textarea, select, details, [tabindex]):not([tabindex="-1"]' is not a valid selector
 ❯ emit node_modules/.deno/nwsapi@2.2.13/node_modules/nwsapi/src/nwsapi.js:575:17
 ❯ Object._matches [as match] node_modules/.deno/nwsapi@2.2.13/node_modules/nwsapi/src/nwsapi.js:1438:9
 ❯ Array.Resolver node_modules/.deno/nwsapi@2.2.13/node_modules/nwsapi/src/nwsapi.js:777:17
 ❯ collect node_modules/.deno/nwsapi@2.2.13/node_modules/nwsapi/src/nwsapi.js:1594:21
 ❯ Object._querySelectorAll [as select] node_modules/.deno/nwsapi@2.2.13/node_modules/nwsapi/src/nwsapi.js:1549:36
 ❯ HTMLDivElementImpl.querySelectorAll node_modules/.deno/jsdom@25.0.1/node_modules/jsdom/lib/jsdom/living/nodes/ParentNode-impl.js:78:26
 ❯ HTMLDivElement.querySelectorAll node_modules/.deno/jsdom@25.0.1/node_modules/jsdom/lib/jsdom/living/generated/Element.js:1119:58
 ❯ HTMLButtonElement.open node_modules/.deno/@skeletonlabs+skeleton@2.10.2/node_modules/@skeletonlabs/skeleton/dist/utilities/Popup/popup.js:119:56
    118|     }
    119|     function close(callback) {
    120|         if (!elemPopup)
       |     ^
    121|             return;
    122|         // Set transition duration
 ❯ HTMLButtonElement.callTheUserObjectsOperation node_modules/.deno/jsdom@25.0.1/node_modules/jsdom/lib/jsdom/living/generated/EventListener.js:26:30
 ❯ innerInvokeEventListeners node_modules/.deno/jsdom@25.0.1/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:350:25

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Serialized Error: { code: 12, INDEX_SIZE_ERR: 1, DOMSTRING_SIZE_ERR: 2, HIERARCHY_REQUEST_ERR: 3, WRONG_DOCUMENT_ERR: 4, INVALID_CHARACTER_ERR: 5, NO_DATA_ALLOWED_ERR: 6, NO_MODIFICATION_ALLOWED_ERR: 7, NOT_FOUND_ERR: 8, NOT_SUPPORTED_ERR: 9, INUSE_ATTRIBUTE_ERR: 10, INVALID_STATE_ERR: 11, SYNTAX_ERR: 12, INVALID_MODIFICATION_ERR: 13, NAMESPACE_ERR: 14, INVALID_ACCESS_ERR: 15, VALIDATION_ERR: 16, TYPE_MISMATCH_ERR: 17, SECURITY_ERR: 18, NETWORK_ERR: 19, ABORT_ERR: 20, URL_MISMATCH_ERR: 21, QUOTA_EXCEEDED_ERR: 22, TIMEOUT_ERR: 23, INVALID_NODE_TYPE_ERR: 24, DATA_CLONE_ERR: 25, __proto__: { constructor: 'Function<DOMException>', name: '<unserializable>', message: '<unserializable>', code: '<unserializable>', INDEX_SIZE_ERR: 1, DOMSTRING_SIZE_ERR: 2, HIERARCHY_REQUEST_ERR: 3, WRONG_DOCUMENT_ERR: 4, INVALID_CHARACTER_ERR: 5, NO_DATA_ALLOWED_ERR: 6, NO_MODIFICATION_ALLOWED_ERR: 7, NOT_FOUND_ERR: 8, NOT_SUPPORTED_ERR: 9, INUSE_ATTRIBUTE_ERR: 10, INVALID_STATE_ERR: 11, SYNTAX_ERR: 12, INVALID_MODIFICATION_ERR: 13, NAMESPACE_ERR: 14, INVALID_ACCESS_ERR: 15, VALIDATION_ERR: 16, TYPE_MISMATCH_ERR: 17, SECURITY_ERR: 18, NETWORK_ERR: 19, ABORT_ERR: 20, URL_MISMATCH_ERR: 21, QUOTA_EXCEEDED_ERR: 22, TIMEOUT_ERR: 23, INVALID_NODE_TYPE_ERR: 24, DATA_CLONE_ERR: 25, toString: 'Function<toString>', __defineGetter__: 'Function<__defineGetter__>', __defineSetter__: 'Function<__defineSetter__>', hasOwnProperty: 'Function<hasOwnProperty>', __lookupGetter__: 'Function<__lookupGetter__>', __lookupSetter__: 'Function<__lookupSetter__>', isPrototypeOf: 'Function<isPrototypeOf>', propertyIsEnumerable: 'Function<propertyIsEnumerable>', valueOf: 'Function<valueOf>', __proto__: { constructor: 'Function<Error>', name: 'Error', message: '', toString: 'Function<toString>', __defineGetter__: 'Function<__defineGetter__>', __defineSetter__: 'Function<__defineSetter__>', hasOwnProperty: 'Function<hasOwnProperty>', __lookupGetter__: 'Function<__lookupGetter__>', __lookupSetter__: 'Function<__lookupSetter__>', isPrototypeOf: 'Function<isPrototypeOf>', propertyIsEnumerable: 'Function<propertyIsEnumerable>', valueOf: 'Function<valueOf>', __proto__: { constructor: 'Function<Object>', __defineGetter__: 'Function<__defineGetter__>', __defineSetter__: 'Function<__defineSetter__>', hasOwnProperty: 'Function<hasOwnProperty>', __lookupGetter__: 'Function<__lookupGetter__>', __lookupSetter__: 'Function<__lookupSetter__>', isPrototypeOf: 'Function<isPrototypeOf>', propertyIsEnumerable: 'Function<propertyIsEnumerable>', toString: 'Function<toString>', valueOf: 'Function<valueOf>', __proto__: null, toLocaleString: 'Function<toLocaleString>' }, toLocaleString: 'Function<toLocaleString>' }, toLocaleString: 'Function<toLocaleString>' } }
This error originated in "src/lib/components/skeleton-test/Tooltip.test.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
The latest test that might've caused the error is "should render the tooltip content on hover". It might mean one of the following:
- The error was thrown, while Vitest was running this test.
- If the error occurred after the test had been completed, this was the last documented test before it was thrown.
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
```

## Reproduction steps

1. Setup SvelteKit: https://svelte.dev/docs/kit/creating-a-project
2. Setup Tailwind: https://tailwindcss.com/docs/guides/sveltekit
3. Setup Vitest + Svelte Testing Library: https://testing-library.com/docs/svelte-testing-library/setup
   * **Note:** Don't install `@sveltejs/vite-plugin-svelte` since it isn't needed for SvelteKit and causes errors
4. Added a workaround for `window.matchMedia`: https://github.com/vitest-dev/vitest/issues/821
5. Install userEvent for simulating user interaction: https://testing-library.com/docs/user-event/install


## Code

### Tooltip.svelte

```svelte
<script lang="ts">
	import { popup, type PopupSettings } from '@skeletonlabs/skeleton';

	export let label: string;
	export let target: string;
	export let content: string;

	const popupHover: PopupSettings = {
		event: 'hover',
		target,
		placement: 'top'
	};
</script>

<button class="btn variant-filled [&>*]:pointer-events-none" use:popup={popupHover}>
	<span>{label}</span>
</button>

<div class="card p-4 variant-filled-secondary" data-popup={target}>
	<p>{content}</p>
	<div class="arrow variant-filled-secondary" />
</div>
```

### TooltipTest.svelte

```svelte
<script lang="ts">
	import { initializeStores, storePopup } from '@skeletonlabs/skeleton';
	import { computePosition, autoUpdate, offset, shift, flip, arrow } from '@floating-ui/dom';
	import Tooltip from './Tooltip.svelte';

	initializeStores();
	storePopup.set({ computePosition, autoUpdate, offset, shift, flip, arrow });
</script>

<Tooltip
	target="info"
	label="More Info"
	content="We're going to check if this component can be tested in Testing Library"
/>
```

###  Tooltip.test.ts

```typescript
import { render } from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';
import TooltipTest from './TooltipTest.svelte';

describe('TooltipTest.svelte', () => {
	it('should render the tooltip content on hover', async () => {
		const { getByText, queryByText } = render(TooltipTest);

		const triggerElement = getByText('More Info');
		await userEvent.hover(triggerElement);

		expect(
			getByText("We're going to check if this component can be tested in Testing Library")
		).toBeVisible();

		await userEvent.unhover(triggerElement);

		expect(
			queryByText("We're going to check if this component can be tested in Testing Library")
		).not.toBeVisible();
	});
});
```

**Reply by endigo9740** [reply, marked as answer]
> So, how should Skeleton be set up for Testing Library?

Skeleton has no special requirements that would require you to use any configuration beyond what Test Library itself requires. If you follow my recommendation above, typically you're just testing components, which is what Vitest with Testing Library is all about.

> So, how should Skeleton be set up for Testing Library?

I'll emphasize again, I can help push you in the right direction (as far as recommendations), but it's outside the scope of our support to teach you how to use Testing Library. I strongly recommend you read through their documentation, look at guides, and understand the plethora of options they provide.

https://testing-library.com/

> By the way, I don't see any test example for [Popup](https://github.com/skeletonlabs/skeleton/tree/dev/packages/skeleton/src/lib/utilities/Popup). Is there any reason for that?

The popup feature historically went through several iterations in Skeleton v1 and v2 and we were never quite satisfied with where it landed. It was never quite a "finished" feature and testing was lacking for some features. We're taking a very different stance in v3, testing will be standardized, and we're changing how [v3 popovers and modals](https://github.com/skeletonlabs/skeleton/discussions/2886) are handled. Ownership of these features will now be directly in developer's hands.

> However, when I try to test it via Testing Library, I get the below error. Seems like it is caused by userEvent.hover.

Perhaps? But I'm not an expert in Testing Library. I'd suggest bringing up this issue in their support community.

**Reply by mcous** [reply]
@endigo9740 howdy, svelte-testing-library maintainer here! Would you mind elaborating on this point - I'm always looking to improve documentation:

> This is due to Vitest and Test Library not having thorough documentation for the Svelte 5 pre-release versions. Now that Svelte 5 has launched, we expect that to change

Svelte 5 shouldn't require any testing setup changes compared to Svelte 4, so I'd love to hear about any issues that you ran into!

(As for the bug mentioned above, it's a CSS parsing bug from a child dependency of JSDom, and not an issue in testing-library nor skeleton. Resolution is to use happy-DOM or a real browser, instead.)

**Reply by endigo9740** [reply]
@mcous Hey Michael, thanks for reaching out. That's strange, I use Vitest + Test Library for testing Svelte 4 components as part of my day job and am fairly well versed in the process. But I definitely recall we had several issues with Svelte 5 components. To the point we abandoned this with plans to revisit as time allowed. Maybe just user error on our end?

Admittedly this has been several months ago, so I can't recall the specifics. But we very much want to bring our Svelte component tests inline with our React components prior to launch. It's one of our final items remaining in fact:

- https://github.com/skeletonlabs/skeleton/issues/2363

We plan to revisit this in the near future (next couple of weeks). Knock on wood everything goes smoothly this time, but would you mind if I reach out directly if we run into similar issues again?

**Reply by mcous** [reply]
Please do! Very happy to help in either GH or Discord. We're embarking on a similar journey at my work so I have a very vested interest in making sure STL behaves smoothly on all versions.

FWIW STL's CI tests Svelte 3, 4, and 5 as well as Jest, Vitest + JSDom, and Vitest + happy-dom. Not to say issues don't exist, but we should be well equipped to fix any such issues without too much fuss!

**Reply by endigo9740** [reply]
Awesome, will do then! We're aiming to launch our first beta version next week, so obviously a lot to juggle at the moment.

But I wouldn't shy away from a preemptive audit if you have a couple minutes to spare. Just make sure we're not doing something dumb. The Skeleton v3 Svelte component package can be found here:
https://github.com/skeletonlabs/skeleton/tree/next/packages/skeleton-svelte

We have a few things in place:

- Several of the required dependencies look to be in `package.json`
- We have a [vitest-setup.js](https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-svelte/vitest-setup.js) - but it's empty, which doesn't seem right
- The [vitest-config.js](https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-svelte/vitest.config.js) is also present with some basic configuration
- And I see we have one [absolute minimal test spec](https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-svelte/src/lib/components/Accordion/Accordion.svelte.spec.ts) for our Accordion component

I know we had wrote our own tests at one point, but I think we opted to strip those out to avoid CI failure until we could debug further.

Again, no pressure, but it might give us a head start this next go round. Just drop any feedback here and I'll see it:

- https://github.com/skeletonlabs/skeleton/issues/2363

**Reply by mcous** [reply]
Quick audit:

- `@testing-library/svelte` needs to be bumped to its latest version (`^5`) for stable Svelte 5 support
    - Probably also want `@sveltejs/vite-plugin-svelte` at its latest stable version, too
- There's both `vitest.config.ts` and `vite.config.ts`, and the `vitest.config.ts` doesn't work
    - Just use `vite.config.ts` to make sure you're running the SK plugin in Vitest
    - Also, you can add the `svelteTesting` plugin from `@testing-library/svelte/vite` rather than messing with `resolve.conditions` yourself
- Empty `vitest-setup.js` is a warning sign, you'll want to add `@testing-library/jest-dom` in there probably according to their setup intructions

From there, it looks like you'll have some test design work to do for various components due to your usage of snippets and contexts - at least in `<Accordion>` - but those concerns aren't different from Svelte 4 children and contexts in tests

- [Slots testing example](https://testing-library.com/docs/svelte-testing-library/example#slots)
    - Should probably be updated with Svelte 5 + snippets example, but is the same exact concept
    - That being said, since snippets come in via props, there may be a different approaches you could take here. I have yet to explore the space myself
- [Context testing example](https://testing-library.com/docs/svelte-testing-library/example#contexts)

**Reply by endigo9740** [reply]
That's super useful Michael, thank you sir! I've linked this post in our ticket to ensure we don't miss this. Much appreciated!

**Reply by endigo9740** [reply]
@mcous just wanted to confirm that I've finally had a chance to circle back to this and your instructions were brilliant. This has worked like a charm. I'll move forward on all our Svelte test cases. Thanks again for the help here!

**Comment by is-sematell** [comment]
For anyone who is also running into this issue, I've opened a ticket in the [testing-library repo](https://github.com/testing-library/svelte-testing-library/issues/406#issuecomment-2437558944)

They have suggested two workarounds, and both of them work:
* Switching the Vitest environment from `jsdom` to [`happy-dom`](https://github.com/capricorn86/happy-dom)
* Use [Vitest's browser mode](https://vitest.dev/guide/browser/)
